### PR TITLE
fix transform_inclusive_scan with different value types

### LINKED ISSUE #1260: transform_inclusive_scan broken with different value types
With previous versions of Thrust, we used to be able to pass in iterators of different value types to `thrust::transform_inclusive_scan` for the transform the inclusive scan stages, with the `UnaryFunction` converting between the two. For example, https://github.com/rapidsai/cudf/blob/branch-0.16/cpp/src/copying/concatenate.cu#L74. But this seems to be broken with the version at head:
```bash
/home/rou/src/cudf/cpp/build/_deps/thrust-src/thrust/iterator/transform_iterator.h(316): error: no suitable constructor exists to convert from "size_t" to "cudf::column_device_view"
          detected during:
            instantiation of "thrust::detail::transform_iterator_base<AdaptableUnaryFunction, Iterator, Reference, Value>::type::reference thrust::transform_iterator<AdaptableUnaryFunction, Iterator, Reference, Value>::dereference() const [with AdaptableUnaryFunction=lambda [](const auto &)->auto, Iterator=thrust::detail::normal_iterator<const cudf::column_device_view *>, Reference=cudf::column_device_view, Value=thrust::use_default]" 
/home/rou/src/cudf/cpp/build/_deps/thrust-src/thrust/iterator/iterator_facade.h(128): here
            instantiation of "Facade::reference thrust::iterator_core_access::dereference(const Facade &) [with Facade=thrust::transform_iterator<lambda [](const auto &)->auto, thrust::detail::normal_iterator<const cudf::column_device_view *>, cudf::column_device_view, thrust::use_default>]" 
/home/rou/src/cudf/cpp/build/_deps/thrust-src/thrust/iterator/iterator_facade.h(310): here
            instantiation of "thrust::iterator_facade<Derived, Value, System, Traversal, Reference, Difference>::reference thrust::iterator_facade<Derived, Value, System, Traversal, Reference, Difference>::operator*() const [with Derived=thrust::transform_iterator<lambda [](const auto &)->auto, thrust::detail::normal_iterator<const cudf::column_device_view *>, cudf::column_device_view, thrust::use_default>, Value=cudf::column_device_view, System=thrust::host_system_tag, Traversal=std::random_access_iterator_tag, Reference=cudf::column_device_view, Difference=std::ptrdiff_t]" 
/home/rou/src/cudf/cpp/build/_deps/thrust-src/thrust/system/detail/sequential/scan.h(67): here
            instantiation of "OutputIterator thrust::system::detail::sequential::inclusive_scan(thrust::system::detail::sequential::execution_policy<DerivedPolicy> &, InputIterator, InputIterator, OutputIterator, BinaryFunction) [with DerivedPolicy=thrust::system::cpp::detail::par_t, InputIterator=thrust::transform_iterator<lambda [](const auto &)->auto, thrust::detail::normal_iterator<const cudf::column_device_view *>, cudf::column_device_view, thrust::use_default>, OutputIterator=thrust::detail::normal_iterator<size_t *>, BinaryFunction=thrust::plus<size_t>]" 
/home/rou/src/cudf/cpp/build/_deps/thrust-src/thrust/detail/scan.inl(63): here
            instantiation of "OutputIterator thrust::inclusive_scan(const thrust::detail::execution_policy_base<DerivedPolicy> &, InputIterator, InputIterator, OutputIterator, AssociativeOperator) [with DerivedPolicy=thrust::system::cpp::detail::par_t, InputIterator=thrust::transform_iterator<lambda [](const auto &)->auto, thrust::detail::normal_iterator<const cudf::column_device_view *>, cudf::column_device_view, thrust::use_default>, OutputIterator=thrust::detail::normal_iterator<size_t *>, AssociativeOperator=thrust::plus<size_t>]" 
/home/rou/src/cudf/cpp/build/_deps/thrust-src/thrust/system/detail/generic/transform_scan.inl(57): here
            instantiation of "OutputIterator thrust::system::detail::generic::transform_inclusive_scan(thrust::execution_policy<ExecutionPolicy> &, InputIterator, InputIterator, OutputIterator, UnaryFunction, BinaryFunction) [with ExecutionPolicy=thrust::system::cpp::detail::par_t, InputIterator=thrust::detail::normal_iterator<const cudf::column_device_view *>, OutputIterator=thrust::detail::normal_iterator<size_t *>, UnaryFunction=lambda [](const auto &)->auto, BinaryFunction=thrust::plus<size_t>]" 
/home/rou/src/cudf/cpp/build/_deps/thrust-src/thrust/detail/transform_scan.inl(47): here
            instantiation of "OutputIterator thrust::transform_inclusive_scan(const thrust::detail::execution_policy_base<DerivedPolicy> &, InputIterator, InputIterator, OutputIterator, UnaryFunction, AssociativeOperator) [with DerivedPolicy=thrust::system::cpp::detail::par_t, InputIterator=thrust::detail::normal_iterator<const cudf::column_device_view *>, OutputIterator=thrust::detail::normal_iterator<size_t *>, UnaryFunction=lambda [](const auto &)->auto, AssociativeOperator=thrust::plus<size_t>]" 
/home/rou/src/cudf/cpp/src/copying/concatenate.cu(74): here
```
Probably shouldn't assume the `InputIterator` and `OutputIterator` would have the same value type.

### PR DESCRIPTION
**Pull request by rongou**
Fixes #1260 

@allisonvacanti @brycelelbach

### CODE REVIEW

File: thrust/system/cuda/detail/transform_scan.h [hunk 1]
```diff
@@ -50,8 +50,8 @@ transform_inclusive_scan(execution_policy<Derived> &policy,
                          TransformOp                transform_op,
                          ScanOp                     scan_op)
 {
-  // Use the input iterator's value type per https://wg21.link/P0571
```
Comment (allisonvacanti) [inline thrust/system/cuda/detail/transform_scan.h:53]: Please restore this comment. It clarifies why we choose certain behavior for consistency in edge cases. Before there was a bit of a mismash of input value types, binary functor results, output iterators, initial value types, etc between the different algorithms and implementations.

File: thrust/system/detail/generic/transform_scan.inl [hunk 2]
```diff
@@ -48,8 +48,8 @@ __host__ __device__
                                           UnaryFunction unary_op,
                                           BinaryFunction binary_op)
 {
-  // Use the input iterator's value type per https://wg21.link/P0571
```
Comment (allisonvacanti) [inline thrust/system/detail/generic/transform_scan.inl:51]: Same for this comment.

File: thrust/system/cuda/detail/transform_scan.h [hunk 3]
```diff
@@ -50,8 +50,8 @@ transform_inclusive_scan(execution_policy<Derived> &policy,
...
-  using result_type = typename thrust::iterator_value<InputIt>::type;
+  using input_type = typename thrust::iterator_value<InputIt>::type;
+  using result_type = typename std::result_of<TransformOp(input_type)>::type;
```
Comment (allisonvacanti) [inline thrust/system/cuda/detail/transform_scan.h:54]: `std::result_of` is deprecated in C++17. Can you update this to use the same pattern as `transform_input_output_iterator`:

https://github.com/thrust/thrust/blob/767647039fc78eccaa643a5d89b3bc863bb48a0f/thrust/iterator/detail/transform_input_output_iterator.inl#L35-L40

File: thrust/system/detail/generic/transform_scan.inl [hunk 4]
```diff
@@ -48,8 +48,8 @@ __host__ __device__
...
-  using ValueType = typename thrust::iterator_value<InputIterator>::type;
+  using InputType = typename thrust::iterator_value<InputIterator>::type;
+  using ValueType = typename std::result_of<UnaryFunction(InputType)>::type;
```
Comment (allisonvacanti) [inline thrust/system/detail/generic/transform_scan.inl:52]: Same note about `result_of`

File: thrust/system/cuda/detail/transform_scan.h (hunk 1 above)
Comment (rongou) [reply thrust/system/cuda/detail/transform_scan.h:53]: Done.

File: thrust/system/cuda/detail/transform_scan.h (hunk 3 above)
Comment (rongou) [reply thrust/system/cuda/detail/transform_scan.h:56]: Done.

File: thrust/system/detail/generic/transform_scan.inl (hunk 2 above)
Comment (rongou) [reply thrust/system/detail/generic/transform_scan.inl:51]: Done.

File: thrust/system/detail/generic/transform_scan.inl (hunk 4 above)
Comment (rongou) [reply thrust/system/detail/generic/transform_scan.inl:54]: Done.

**Comment by rongou** [comment]
Forgot to mention, `transform_exclusive_scan` doesn't have this issue since we have to pass in the `InitialValueType`, which is the same as the result type of the transformation operation.

**Comment by allisonvacanti** [comment]
LGTM.

CI running under 28993675.

**Comment by allisonvacanti** [comment]
Restarting CI, 3 DVS builders failed spuriously in the last batch.

Same CL.

**Comment by allisonvacanti** [comment]
Looks good -- merging.

**Pull request merged**